src/librustc/infer/error_reporting/mod.rs

```diff
@@ -798,7 +798,7 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                 if !has_default {
                     break;
                 }
-                if self.tcx.type_of(def_id).subst(self.tcx, substs) != actual {
+                if self.tcx.type_of(def_id).subst(self.tcx, &substs) != actual {
```

Comment:
I dislike all these additional references. Can we get rid of those via some trait impls or by changing the arguments of all the functions?